torch/csrc/distributed/c10d/ProcessGroupGloo.cpp

```diff
@@ -1661,7 +1661,7 @@ class AsyncAllgatherCoalescedWork : public ProcessGroupGloo::AsyncWork {
     const auto& scalarType = input_list[0].scalar_type();
     gloo::AllgatherOptions opts(context_);
     opts.setTag(tag);
-    opts.setTimeout(timeout_);
```

Comment:
i'm not familiar with gloo's PG design. Do you have an explanation for what `timeout_` represents?  is it dead code?  Should we delete it?  

cc @d4l3k @fduwjj

Reply:
`timeout_` is not dead code.. it's essential for per-operation timeout overrides. It has two usecases: process-group timeout (what we're fixing)

```
dist._set_pg_timeout(timedelta(seconds=1), pg)  # sets context timeout
dist.all_reduce(tensor)  # uses 1 second
```

and, per-operation override (existing feature)

```
opts = dist.AllreduceOptions()
opts.timeout = timedelta(milliseconds=100)  # override for this op only
pg.allreduce([tensor], opts)  # uses 100ms, not process-group timeout
```

so here `timeout_` stores the per-operation override (or `kUnsetTimeout` if none) and `getTimeout()` returns `timeout_` if set, otherwise queries `context_->getTimeout()` dynamically